- Name of activity: Search with Numpy and Pandas
- Your name: Dima Mikhaylov
- Your UVA computing IDs: agp7dp

## 1. Exploratory Data Analysis 

In [1]:
# import pandas and inspect get column names
import pandas as pd
df = pd.read_csv('weather.csv', index_col=0)
df.head(3)

Date Time  p (mbar)  T (degC)  rho (g/m**3)
0  01.01.2009 00:10:00    996.52     -8.02       1307.75
1  01.01.2009 00:20:00    996.57     -8.41       1309.80
2  01.01.2009 00:30:00    996.53     -8.51       1310.24

In [2]:
# Convert date to_datetime type and separate year into a new column
df["AsOfDate"] = pd.to_datetime(df['Date Time'])
df["AsOfYear"] = df.AsOfDate.dt.year
df.head()

Date Time  p (mbar)  T (degC)  rho (g/m**3)            AsOfDate  \
0  01.01.2009 00:10:00    996.52     -8.02       1307.75 2009-01-01 00:10:00   
1  01.01.2009 00:20:00    996.57     -8.41       1309.80 2009-01-01 00:20:00   
2  01.01.2009 00:30:00    996.53     -8.51       1310.24 2009-01-01 00:30:00   
3  01.01.2009 00:40:00    996.51     -8.31       1309.19 2009-01-01 00:40:00   
4  01.01.2009 00:50:00    996.51     -8.27       1309.00 2009-01-01 00:50:00   

   AsOfYear  
0      2009  
1      2009  
2      2009  
3      2009  
4      2009

## 2. Preliminary results

TODO:
- What was the average temperature in 2009?
- What was the highest temperature in 2010? When did it happen?
- Which year has the lowest average temperature?

In [3]:
# Group by temp by year and use mean() for aggregation
df["T (degC)"].groupby(df.AsOfYear).mean()

AsOfYear
2009     8.830284
2010     7.504652
2011     9.303914
2012     9.655535
2013     9.093978
2014    10.718917
2015    10.511207
2016     9.988671
2017    -4.820000
Name: T (degC), dtype: float64

#### Observation 1: the average temperature in 2009 was 8.83

In [4]:
# Take all temp and date records for 2010, use max() to aggregate
df[["T (degC)", "Date Time"]].loc[df.AsOfYear==2010].max()

T (degC)                   34.92
Date Time    31.12.2010 23:50:00
dtype: object

#### Observation 2: the highest temperature in 2010 was almost 35 degrees.

In [5]:
df[["T (degC)"]].loc[df.AsOfYear==2017].mean()

T (degC)   -4.82
dtype: float64

In [6]:
df.tail()

Date Time  p (mbar)  T (degC)  rho (g/m**3)  \
420546  31.12.2016 23:20:00   1000.07     -4.05       1292.98   
420547  31.12.2016 23:30:00    999.93     -3.35       1289.44   
420548  31.12.2016 23:40:00    999.82     -3.16       1288.39   
420549  31.12.2016 23:50:00    999.81     -4.23       1293.56   
420550  01.01.2017 00:00:00    999.82     -4.82       1296.38   

                  AsOfDate  AsOfYear  
420546 2016-12-31 23:20:00      2016  
420547 2016-12-31 23:30:00      2016  
420548 2016-12-31 23:40:00      2016  
420549 2016-12-31 23:50:00      2016  
420550 2017-01-01 00:00:00      2017

In [39]:
df["T (degC)"].groupby(df.AsOfYear).mean()

AsOfYear
2009     8.830284
2010     7.504652
2011     9.303914
2012     9.655535
2013     9.093978
2014    10.718917
2015    10.511207
2016     9.988671
2017    -4.820000
Name: T (degC), dtype: float64

#### Observation 3: surprisingly, 2017 has average temperature of only -4.8 degrees what is due to the lack of meaningful data, i.e. only one observation 420550. Therefore, a better answer for the lowest average temperature is probably 2010 with 7.5

## 3. Solutions template

In [149]:
def average(DataFrame, col_name, year):
    
    """ 
    Takes a DataFrame object, column name, and a year
    Returns list average by column for the year."""

    avg = DataFrame[col_name].groupby(df.AsOfYear).mean()
    return [avg[year]]

def search_high(DataFrame, col_name, col_datetime, year):
    
    """
    Takes a DataFrame object, specific value column, datetime and a year for aggregation.
    Returns highest annual value and the datetime when it was recorded."""
 
    highest = DataFrame[[col_name, col_datetime]].loc[df.AsOfYear==year].max()
    return [highest]

def search_lowest_year(DataFrame, column, check_df = True):
    """
    Takes a DataFrame object, specific value column, and check_df to ensure averaging.
    Returns the year with the lowest average temperature."""
    
    df=0
    if check_df:
        df=2 # Min 2 observations are needed to calculate mean
        grouped = c_df.groupby('AsOfYear').filter(lambda x: x[column].count() > df)
        
    else: # alternatively, df is set to 0, effectively this check is turned off
        grouped = c_df.groupby('AsOfYear').filter(lambda x: x[column].count() > df)
    
    lowest_year = grouped[["T (degC)","AsOfYear"]].groupby( #take temp and year
        grouped.AsOfYear).mean().sort_values('T (degC)').iloc[0] # group by year and sort by temp
    
    return lowest_year

## 4. Testing 

In [94]:
# Report the average temperature for 2009:
column = "T (degC)"
year = 2009
average(df, column, year)

[8.830283501721938]

In [95]:
# What was the highest temperature in 2010? When did it happen?
column = "T (degC)"
when = "Date Time"
year = 2010
search_high(df, column, when, year)

[T (degC)                   34.92
 Date Time    31.12.2010 23:50:00
 dtype: object]

In [156]:
# Which year has the lowest average temperature?
column = "T (degC)"
search_lowest_year(df, column)


T (degC)       7.504652
AsOfYear    2010.000000
Name: 2010, dtype: float64